src/passes/OptimizeCasts.cpp

```diff
@@ -114,22 +117,50 @@ struct BestCastFinder : public LinearExecutionWalker<BestCastFinder> {
 
   static void doNoteNonLinear(BestCastFinder* self, Expression** currp) {
     self->mostCastedGets.clear();
+    self->curCastedSets.clear();
   }
 
   void visitLocalSet(LocalSet* curr) {
     // Clear any information about this local; it has a new value here.
     mostCastedGets.erase(curr->index);
+
+    // This only checks the immediate child for casts. This should be extended
+    // to look deeper for casts
+    if (curr->value->dynCast<RefAs>() || curr->value->dynCast<RefCast>()) {
+      curCastedSets[curr->index] = curr->value;
+    } else {
+      // If the local.set doesn't use a cast, get rid of any old cast information
+      curCastedSets.erase(curr->index);
+    }
   }
 
   void visitLocalGet(LocalGet* curr) {
-    auto iter = mostCastedGets.find(curr->index);
-    if (iter != mostCastedGets.end()) {
-      auto* bestCast = iter->second;
+    auto getIter = mostCastedGets.find(curr->index);
+    auto setIter = curCastedSets.find(curr->index);
+
+    if (getIter != mostCastedGets.end()) {
+      auto* bestCast = getIter->second;
+      if (setIter != curCastedSets.end()) {
+        // Always use a cast in local.set if it is equal or better than
+        // a local.get since we know it is always before any gets that
+        // retrieve the set value from the index
+        if (bestCast->type == setIter->second->type ||
+            Type::isSubType(setIter->second->type, bestCast->type)) {
```

Comment:
Subtyping includes equality, so there's no need to separate check equality in the code. I do think it's helpful for the reader that you call out "equal or better" in the comment, though.

```suggestion
        if (Type::isSubType(setIter->second->type, bestCast->type)) {
```